Algorithm.Python/Alphas/ContingentClaimsAnalysisDefaultPredictionAlpha.py

```diff
@@ -0,0 +1,266 @@
+                                                    (x.FinancialStatements.BalanceSheet.CurrentLiabilities.TwelveMonths > 0) and
+                                                    (x.FinancialStatements.BalanceSheet.InterestPayable.TwelveMonths > 0) and
+                                                    (x.OperationRatios.TotalAssetsGrowth.OneYear > 0) and
+                                                    (x.FinancialStatements.IncomeStatement.GrossDividendPayment.TwelveMonths > 0) and
+                                                    (x.OperationRatios.ROA.OneYear > 0), reverse=True)
+
+            self.symbols = [ x.Symbol for x in fineFilter ]
+
+            return self.symbols
+
+
+class ContingentClaimsAnalysisAlphaModel:
+    
+    def __init__(self, *args, **kwargs):
+        self.symbolDataBySymbol = {}
+        self.month = None
+        self.default_threshold = kwargs['default_threshold'] if 'default_threshold' in kwargs else 0.25
+        self.expiry = None
+        self.start_time = kwargs['start_time'] if 'start_time' in kwargs else None
+        self.epsilon = kwargs['epsilon'] if 'epsilon' in kwargs else 0.00001     ## This serves as a check to filter out symbols with a default probability of, e.g., 2.89e-20
+        
+
+    ## This model applies options pricing theory, Black-Scholes specifically, to fundamental data
+    ## to give the probability of a default
+    def CCADefaultProbability(self, algorithm, symbolData):
+        
+        B = symbolData.DebtValue
+        V = symbolData.AssetValue
+        D = symbolData.DividendAndInterest
+        sigma = symbolData.ValuationChangeVolatility
+        tau = symbolData.TimeToMaturity - ((self.start_time - algorithm.Time).days)
+        mu = symbolData.ROA
+        
+        d2 = ((np.log(V) - np.log(B)) + ((mu - D) - 0.5*sigma**2.0)*tau)/ (sigma*np.sqrt(tau))
+        probability_of_default = sp.norm.cdf(-d2)
+            
+        return probability_of_default
+
+
+    def Update(self, algorithm, data):
+        
+        '''Updates this alpha model with the latest data from the algorithm.
+        This is called each time the algorithm receives data for subscribed securities
+        Args:
+            algorithm: The algorithm instance
+            data: The new data available
+        Returns:
+            The new insights generated'''
+            
+        ## Build a list to hold our insights
+        insights = []
+        
+        for symbol, symbolData in self.symbolDataBySymbol.items():
+            pod = self.CCADefaultProbability(algorithm, symbolData)
+            algorithm.Log('POB for ' + str(symbol) + ': ' + str(pod))
+            ## If Prob. of Default is greater than our set threshold, then emit an insight indicating that this asset is trending downward
+            if (pod >= self.default_threshold) and (pod != 1.0):
+                insights.append(Insight(symbol, timedelta(days = 30), InsightType.Price, InsightDirection.Down, pod, None))
```

Comment:
Just a comment, not a required change:
We can use the static helper method `Insight.Price` and the insight period could be configurable:
```python
insights.append(Insight.Price(symbol, self.period, InsightDirection.Down, pod, None))
```
where `self.period` is defined in `__init__` as `self.period = timedelta(days = 30)`.